commit 383bbb2771f36a74b1690c08fb2bdd0921695091
Author: Susan Franklin <sfranklin>
Date:   Tue Jan 29 19:35:14 2008 +0000

    Bug 214782 - [prov] [ui] - repository properties dialog should not load the repo

1	1	bundles/org.eclipse.equinox.p2.ui.admin/src/org/eclipse/equinox/internal/p2/ui/admin/ArtifactRepositoriesView.java
1	1	bundles/org.eclipse.equinox.p2.ui.admin/src/org/eclipse/equinox/internal/p2/ui/admin/MetadataRepositoriesView.java
1	1	bundles/org.eclipse.equinox.p2.ui.admin/src/org/eclipse/equinox/internal/p2/ui/admin/RepositoriesView.java
0	2	bundles/org.eclipse.equinox.p2.ui.admin/src/org/eclipse/equinox/internal/p2/ui/admin/dialogs/ProfileGroup.java
4	22	bundles/org.eclipse.equinox.p2.ui.admin/src/org/eclipse/equinox/internal/p2/ui/admin/dialogs/ProfilePropertyPage.java
2	2	bundles/org.eclipse.equinox.p2.ui.admin/src/org/eclipse/equinox/internal/p2/ui/admin/dialogs/RepositoryImplementationPropertyPage.java
1	1	bundles/org.eclipse.equinox.p2.ui.sdk/plugin.xml
1	1	bundles/org.eclipse.equinox.p2.ui.sdk/src/org/eclipse/equinox/internal/p2/ui/sdk/RepositoryManipulationDialog.java
4	2	bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/ProvUIMessages.java
3	2	bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/messages.properties
26	6	bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/model/ArtifactRepositoryElement.java
1	1	bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/p2/ui/dialogs/AddRepositoryDialog.java
33	46	bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/p2/ui/dialogs/RepositoryPropertyPage.java
36	5	bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/p2/ui/model/MetadataRepositoryElement.java
7	1	bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/p2/ui/model/RepositoryElement.java
4	4	bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/p2/ui/operations/ProvisioningUtil.java
1	1	bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/p2/ui/viewers/ProvElementComparer.java
diff --git a/bundles/org.eclipse.equinox.p2.ui.admin/src/org/eclipse/equinox/internal/p2/ui/admin/ArtifactRepositoriesView.java b/bundles/org.eclipse.equinox.p2.ui.admin/src/org/eclipse/equinox/internal/p2/ui/admin/ArtifactRepositoriesView.java
index 431d0a6..286975a 100644
--- a/bundles/org.eclipse.equinox.p2.ui.admin/src/org/eclipse/equinox/internal/p2/ui/admin/ArtifactRepositoriesView.java
+++ b/bundles/org.eclipse.equinox.p2.ui.admin/src/org/eclipse/equinox/internal/p2/ui/admin/ArtifactRepositoriesView.java
@@ -58,7 +58,7 @@ public class ArtifactRepositoriesView extends RepositoriesView {
 		ArrayList urls = new ArrayList();
 		for (int i = 0; i < elements.length; i++) {
 			if (elements[i] instanceof RepositoryElement)
-				urls.add(((RepositoryElement) elements[i]).getURL());
+				urls.add(((RepositoryElement) elements[i]).getLocation());
 		}
 		return new RemoveArtifactRepositoryOperation(ProvAdminUIMessages.ArtifactRepositoriesView_RemoveRepositoryOperationLabel, (URL[]) urls.toArray(new URL[urls.size()]));
 	}
diff --git a/bundles/org.eclipse.equinox.p2.ui.admin/src/org/eclipse/equinox/internal/p2/ui/admin/MetadataRepositoriesView.java b/bundles/org.eclipse.equinox.p2.ui.admin/src/org/eclipse/equinox/internal/p2/ui/admin/MetadataRepositoriesView.java
index b8e6798..9fd329c 100644
--- a/bundles/org.eclipse.equinox.p2.ui.admin/src/org/eclipse/equinox/internal/p2/ui/admin/MetadataRepositoriesView.java
+++ b/bundles/org.eclipse.equinox.p2.ui.admin/src/org/eclipse/equinox/internal/p2/ui/admin/MetadataRepositoriesView.java
@@ -74,7 +74,7 @@ public class MetadataRepositoriesView extends RepositoriesView {
 		ArrayList urls = new ArrayList();
 		for (int i = 0; i < elements.length; i++) {
 			if (elements[i] instanceof RepositoryElement)
-				urls.add(((RepositoryElement) elements[i]).getURL());
+				urls.add(((RepositoryElement) elements[i]).getLocation());
 		}
 		return new RemoveMetadataRepositoryOperation(ProvAdminUIMessages.ArtifactRepositoriesView_RemoveRepositoryOperationLabel, (URL[]) urls.toArray(new URL[urls.size()]));
 	}
diff --git a/bundles/org.eclipse.equinox.p2.ui.admin/src/org/eclipse/equinox/internal/p2/ui/admin/RepositoriesView.java b/bundles/org.eclipse.equinox.p2.ui.admin/src/org/eclipse/equinox/internal/p2/ui/admin/RepositoriesView.java
index 5072e7e..1695055 100644
--- a/bundles/org.eclipse.equinox.p2.ui.admin/src/org/eclipse/equinox/internal/p2/ui/admin/RepositoriesView.java
+++ b/bundles/org.eclipse.equinox.p2.ui.admin/src/org/eclipse/equinox/internal/p2/ui/admin/RepositoriesView.java
@@ -64,7 +64,7 @@ abstract class RepositoriesView extends ProvView {
 			ArrayList urls = new ArrayList();
 			for (int i = 0; i < elements.length; i++)
 				if (elements[i] instanceof RepositoryElement)
-					urls.add(((RepositoryElement) elements[i]).getURL());
+					urls.add(((RepositoryElement) elements[i]).getLocation());
 			openAddRepositoryDialog(getShell(), (URL[]) urls.toArray(new URL[urls.size()]));
 		}
 	}
diff --git a/bundles/org.eclipse.equinox.p2.ui.admin/src/org/eclipse/equinox/internal/p2/ui/admin/dialogs/ProfileGroup.java b/bundles/org.eclipse.equinox.p2.ui.admin/src/org/eclipse/equinox/internal/p2/ui/admin/dialogs/ProfileGroup.java
index d1e2d88..6b9d491 100644
--- a/bundles/org.eclipse.equinox.p2.ui.admin/src/org/eclipse/equinox/internal/p2/ui/admin/dialogs/ProfileGroup.java
+++ b/bundles/org.eclipse.equinox.p2.ui.admin/src/org/eclipse/equinox/internal/p2/ui/admin/dialogs/ProfileGroup.java
@@ -187,8 +187,6 @@ public class ProfileGroup {
 	}
 
 	public void updateProfile() {
-		// We forced the id field to have content so don't 
-		// check its length
 		if (profile == null) {
 			Map profileProperties = new HashMap();
 
diff --git a/bundles/org.eclipse.equinox.p2.ui.admin/src/org/eclipse/equinox/internal/p2/ui/admin/dialogs/ProfilePropertyPage.java b/bundles/org.eclipse.equinox.p2.ui.admin/src/org/eclipse/equinox/internal/p2/ui/admin/dialogs/ProfilePropertyPage.java
index 23a2e0a..c71279d 100644
--- a/bundles/org.eclipse.equinox.p2.ui.admin/src/org/eclipse/equinox/internal/p2/ui/admin/dialogs/ProfilePropertyPage.java
+++ b/bundles/org.eclipse.equinox.p2.ui.admin/src/org/eclipse/equinox/internal/p2/ui/admin/dialogs/ProfilePropertyPage.java
@@ -10,14 +10,11 @@
  *******************************************************************************/
 package org.eclipse.equinox.internal.p2.ui.admin.dialogs;
 
-import org.eclipse.core.runtime.IStatus;
 import org.eclipse.equinox.internal.p2.ui.admin.ProvAdminUIMessages;
 import org.eclipse.equinox.p2.engine.Profile;
 import org.eclipse.equinox.p2.ui.ProvUI;
 import org.eclipse.jface.dialogs.Dialog;
 import org.eclipse.swt.SWT;
-import org.eclipse.swt.events.ModifyEvent;
-import org.eclipse.swt.events.ModifyListener;
 import org.eclipse.swt.widgets.*;
 import org.eclipse.ui.dialogs.PropertyPage;
 
@@ -36,26 +33,11 @@ public class ProfilePropertyPage extends PropertyPage {
 			Label label = new Label(parent, SWT.DEFAULT);
 			label.setText(ProvAdminUIMessages.No_Property_Item_Selected);
 		}
-		profileGroup = new ProfileGroup(parent, profile, new ModifyListener() {
-			public void modifyText(ModifyEvent event) {
-				verifyComplete();
-			}
-		});
+		// Assume that we do not edit profile properties for now
+		noDefaultAndApplyButton();
+
+		profileGroup = new ProfileGroup(parent, profile, null);
 		Dialog.applyDialogFont(profileGroup.getComposite());
-		verifyComplete();
 		return profileGroup.getComposite();
 	}
-
-	void verifyComplete() {
-		if (profileGroup == null) {
-			return;
-		}
-		IStatus status = profileGroup.verify();
-		setValid(status.isOK());
-	}
-
-	public boolean performOk() {
-		profileGroup.updateProfile();
-		return true;
-	}
 }
diff --git a/bundles/org.eclipse.equinox.p2.ui.admin/src/org/eclipse/equinox/internal/p2/ui/admin/dialogs/RepositoryImplementationPropertyPage.java b/bundles/org.eclipse.equinox.p2.ui.admin/src/org/eclipse/equinox/internal/p2/ui/admin/dialogs/RepositoryImplementationPropertyPage.java
index 5cbf7f8..d76dc85 100644
--- a/bundles/org.eclipse.equinox.p2.ui.admin/src/org/eclipse/equinox/internal/p2/ui/admin/dialogs/RepositoryImplementationPropertyPage.java
+++ b/bundles/org.eclipse.equinox.p2.ui.admin/src/org/eclipse/equinox/internal/p2/ui/admin/dialogs/RepositoryImplementationPropertyPage.java
@@ -57,8 +57,8 @@ public class RepositoryImplementationPropertyPage extends RepositoryPropertyPage
 	}
 
 	private void initializeTable() {
-		if (getRepository() != null) {
-			Map repoProperties = getRepository().getProperties();
+		if (getRepositoryElement() != null) {
+			Map repoProperties = getRepositoryElement().getRepository(null).getProperties();
 			if (repoProperties != null) {
 				String[] propNames = (String[]) repoProperties.keySet().toArray(new String[repoProperties.size()]);
 				for (int i = 0; i < propNames.length; i++) {
diff --git a/bundles/org.eclipse.equinox.p2.ui.sdk/plugin.xml b/bundles/org.eclipse.equinox.p2.ui.sdk/plugin.xml
index c01052d..6903547 100644
--- a/bundles/org.eclipse.equinox.p2.ui.sdk/plugin.xml
+++ b/bundles/org.eclipse.equinox.p2.ui.sdk/plugin.xml
@@ -29,7 +29,7 @@
             id="org.eclipse.equinox.internal.p2.ui.sdk.RepositoryPropertyPage">
             <enabledWhen>
             <adapt
-                  type="org.eclipse.equinox.p2.core.repository.IRepository">
+                  type="org.eclipse.equinox.p2.ui.model.RepositoryElement">
              </adapt>
           </enabledWhen>
       </page>
diff --git a/bundles/org.eclipse.equinox.p2.ui.sdk/src/org/eclipse/equinox/internal/p2/ui/sdk/RepositoryManipulationDialog.java b/bundles/org.eclipse.equinox.p2.ui.sdk/src/org/eclipse/equinox/internal/p2/ui/sdk/RepositoryManipulationDialog.java
index 84d1178..d4a611a 100644
--- a/bundles/org.eclipse.equinox.p2.ui.sdk/src/org/eclipse/equinox/internal/p2/ui/sdk/RepositoryManipulationDialog.java
+++ b/bundles/org.eclipse.equinox.p2.ui.sdk/src/org/eclipse/equinox/internal/p2/ui/sdk/RepositoryManipulationDialog.java
@@ -199,7 +199,7 @@ public class RepositoryManipulationDialog extends TrayDialog {
 				List urls = new ArrayList();
 				for (int i = 0; i < selection.length; i++) {
 					if (selection[i] instanceof MetadataRepositoryElement)
-						urls.add(((MetadataRepositoryElement) selection[i]).getURL());
+						urls.add(((MetadataRepositoryElement) selection[i]).getLocation());
 				}
 				if (urls.size() > 0) {
 					RemoveColocatedRepositoryOperation op = new RemoveColocatedRepositoryOperation(ProvSDKMessages.RepositoryManipulationDialog_RemoveOperationLabel, (URL[]) urls.toArray(new URL[urls.size()]));
diff --git a/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/ProvUIMessages.java b/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/ProvUIMessages.java
index a7477bc..b514109 100644
--- a/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/ProvUIMessages.java
+++ b/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/ProvUIMessages.java
@@ -101,6 +101,10 @@ public class ProvUIMessages extends NLS {
 
 	// Property pages
 	public static String IUPropertyPage_NoIUSelected;
+	public static String RepositoryPropertyPage_DescriptionFieldLabel;
+	public static String RepositoryPropertyPage_NameFieldLabel;
+	public static String RepositoryPropertyPage_URLFieldLabel;
+
 	public static String RepositoryPropertyPage_NoRepoSelected;
 
 	// Dialog groups
@@ -108,9 +112,7 @@ public class ProvUIMessages extends NLS {
 	public static String RepositoryGroup_ArchivedRepoBrowseButton;
 	public static String RepositoryGroup_RepositoryFile;
 	public static String RepositoryGroup_SelectRepositoryDirectory;
-	public static String RepositoryGroup_RepositoryNameFieldLabel;
 	public static String RepositoryGroup_URLRequired;
-	public static String RepositoryGroup_RepositoryURLFieldLabel;
 
 	// Dialogs
 	public static String AddRepositoryDialog_DuplicateURL;
diff --git a/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/messages.properties b/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/messages.properties
index 58e9654..c55c1a5 100644
--- a/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/messages.properties
+++ b/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/messages.properties
@@ -65,10 +65,11 @@ IUDetailsLabelProvider_Unknown=Unknown
 IUPropertiesGroup_NameProperty=Name
 IUPropertiesGroup_ProviderProperty=Provider
 IUPropertiesGroup_LicenseProperty=License
+RepositoryPropertyPage_DescriptionFieldLabel=Description:
 RepositoryPropertyPage_NoRepoSelected=There was no repository selected
+RepositoryPropertyPage_NameFieldLabel=Name:
+RepositoryPropertyPage_URLFieldLabel=Location:
 
-RepositoryGroup_RepositoryNameFieldLabel=&Name:
-RepositoryGroup_RepositoryURLFieldLabel=&URL:
 RepositoryGroup_LocalRepoBrowseButton=&Local...
 RepositoryGroup_ArchivedRepoBrowseButton=&Archive...
 RepositoryGroup_SelectRepositoryDirectory=Select a repository root directory:
diff --git a/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/model/ArtifactRepositoryElement.java b/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/model/ArtifactRepositoryElement.java
index ad7b0e9..8072314 100644
--- a/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/model/ArtifactRepositoryElement.java
+++ b/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/model/ArtifactRepositoryElement.java
@@ -15,6 +15,7 @@ import org.eclipse.core.runtime.IProgressMonitor;
 import org.eclipse.core.runtime.jobs.ISchedulingRule;
 import org.eclipse.equinox.internal.p2.ui.ProvUIMessages;
 import org.eclipse.equinox.p2.artifact.repository.IArtifactRepository;
+import org.eclipse.equinox.p2.artifact.repository.IArtifactRepositoryManager;
 import org.eclipse.equinox.p2.core.ProvisionException;
 import org.eclipse.equinox.p2.core.repository.IRepository;
 import org.eclipse.equinox.p2.metadata.IArtifactKey;
@@ -53,7 +54,7 @@ public class ArtifactRepositoryElement extends ProvElement implements IDeferredW
 	}
 
 	protected Object[] fetchChildren(Object o, IProgressMonitor monitor) {
-		IArtifactRepository repository = getRepository(monitor);
+		IArtifactRepository repository = (IArtifactRepository) getRepository(monitor);
 		if (repository == null)
 			return new ArtifactElement[0];
 		IArtifactKey[] keys = repository.getArtifactKeys();
@@ -69,10 +70,10 @@ public class ArtifactRepositoryElement extends ProvElement implements IDeferredW
 		if (name != null && name.length() > 0) {
 			return name;
 		}
-		return getURL().toExternalForm();
+		return getLocation().toExternalForm();
 	}
 
-	private IArtifactRepository getRepository(IProgressMonitor monitor) {
+	public IRepository getRepository(IProgressMonitor monitor) {
 		if (repo == null)
 			try {
 				repo = ProvisioningUtil.loadArtifactRepository(url, monitor);
@@ -105,16 +106,35 @@ public class ArtifactRepositoryElement extends ProvElement implements IDeferredW
 	/* (non-Javadoc)
 	 * @see org.eclipse.equinox.p2.ui.model.RepositoryElement#getURL()
 	 */
-	public URL getURL() {
+	public URL getLocation() {
 		return url;
 	}
 
-	/* (non-Javadoc)
+	/*
+	 * (non-Javadoc)
 	 * @see org.eclipse.equinox.p2.ui.model.RepositoryElement#getName()
 	 */
 	public String getName() {
 		try {
-			return ProvisioningUtil.getArtifactRepositoryName(url);
+			String name = ProvisioningUtil.getArtifactRepositoryProperty(url, IArtifactRepositoryManager.PROP_NAME);
+			if (name == null)
+				return ""; //$NON-NLS-1$
+			return name;
+		} catch (ProvisionException e) {
+			return ""; //$NON-NLS-1$
+		}
+	}
+
+	/*
+	 * (non-Javadoc)
+	 * @see org.eclipse.equinox.p2.ui.model.RepositoryElement#getDescription()
+	 */
+	public String getDescription() {
+		try {
+			String description = ProvisioningUtil.getArtifactRepositoryProperty(url, IArtifactRepositoryManager.PROP_DESCRIPTION);
+			if (description == null)
+				return ""; //$NON-NLS-1$
+			return description;
 		} catch (ProvisionException e) {
 			return ""; //$NON-NLS-1$
 		}
diff --git a/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/p2/ui/dialogs/AddRepositoryDialog.java b/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/p2/ui/dialogs/AddRepositoryDialog.java
index 913849b..b186000 100644
--- a/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/p2/ui/dialogs/AddRepositoryDialog.java
+++ b/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/p2/ui/dialogs/AddRepositoryDialog.java
@@ -66,7 +66,7 @@ public abstract class AddRepositoryDialog extends StatusDialog {
 		comp.setLayoutData(data);
 
 		Label urlLabel = new Label(comp, SWT.NONE);
-		urlLabel.setText(ProvUIMessages.RepositoryGroup_RepositoryURLFieldLabel);
+		urlLabel.setText(ProvUIMessages.RepositoryPropertyPage_URLFieldLabel);
 		url = new Text(comp, SWT.BORDER);
 		data = new GridData(GridData.FILL_HORIZONTAL);
 		data.widthHint = convertHorizontalDLUsToPixels(IDialogConstants.ENTRY_FIELD_WIDTH);
diff --git a/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/p2/ui/dialogs/RepositoryPropertyPage.java b/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/p2/ui/dialogs/RepositoryPropertyPage.java
index 6e8800f..5e95670 100644
--- a/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/p2/ui/dialogs/RepositoryPropertyPage.java
+++ b/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/p2/ui/dialogs/RepositoryPropertyPage.java
@@ -11,8 +11,8 @@
 package org.eclipse.equinox.p2.ui.dialogs;
 
 import org.eclipse.equinox.internal.p2.ui.ProvUIMessages;
-import org.eclipse.equinox.p2.core.repository.IRepository;
 import org.eclipse.equinox.p2.ui.ProvUI;
+import org.eclipse.equinox.p2.ui.model.RepositoryElement;
 import org.eclipse.jface.dialogs.Dialog;
 import org.eclipse.swt.SWT;
 import org.eclipse.swt.layout.GridData;
@@ -27,23 +27,20 @@ import org.eclipse.ui.dialogs.PropertyPage;
  */
 public class RepositoryPropertyPage extends PropertyPage {
 
-	private IRepository repository;
+	private RepositoryElement repositoryElement;
 	private Composite composite;
 	private Text name;
 	private Text url;
+	private Text description;
 
 	protected Control createContents(Composite parent) {
-		// TODO if all we are ever going to show is the name, URL, and
-		// description, this could be done without loading the repo.
-		this.repository = getRepository();
-		if (repository == null) {
+		this.repositoryElement = getRepositoryElement();
+		if (repositoryElement == null) {
 			Label label = new Label(parent, SWT.DEFAULT);
 			label.setText(ProvUIMessages.RepositoryPropertyPage_NoRepoSelected);
 			return label;
 		}
-		if (!repository.isModifiable()) {
-			noDefaultAndApplyButton();
-		}
+		noDefaultAndApplyButton();
 
 		composite = new Composite(parent, SWT.NONE);
 		GridLayout layout = new GridLayout();
@@ -53,55 +50,45 @@ public class RepositoryPropertyPage extends PropertyPage {
 		data.widthHint = 350;
 		composite.setLayoutData(data);
 
+		Label urlLabel = new Label(composite, SWT.NONE);
+		urlLabel.setText(ProvUIMessages.RepositoryPropertyPage_URLFieldLabel);
+		url = new Text(composite, SWT.WRAP | SWT.READ_ONLY);
+		url.setLayoutData(new GridData(GridData.FILL_HORIZONTAL));
+
 		Label nameLabel = new Label(composite, SWT.NONE);
-		nameLabel.setText(ProvUIMessages.RepositoryGroup_RepositoryNameFieldLabel);
+		nameLabel.setText(ProvUIMessages.RepositoryPropertyPage_NameFieldLabel);
+		name = new Text(composite, SWT.WRAP | SWT.READ_ONLY);
+		name.setLayoutData(new GridData(GridData.FILL_HORIZONTAL));
 
-		name = new Text(composite, SWT.BORDER);
+		Label descriptionLabel = new Label(composite, SWT.NONE);
+		descriptionLabel.setText(ProvUIMessages.RepositoryPropertyPage_DescriptionFieldLabel);
+		data = new GridData();
+		data.verticalAlignment = SWT.TOP;
+		descriptionLabel.setLayoutData(data);
+		description = new Text(composite, SWT.WRAP | SWT.READ_ONLY);
 		data = new GridData(GridData.FILL_HORIZONTAL);
-		name.setLayoutData(data);
-		name.setEditable(repository.isModifiable());
-
-		Label urlLabel = new Label(composite, SWT.NONE);
-		urlLabel.setText(ProvUIMessages.RepositoryGroup_RepositoryURLFieldLabel);
-		url = new Text(composite, SWT.BORDER);
-		url.setLayoutData(new GridData(GridData.FILL_HORIZONTAL));
-		url.setEditable(false);
+		data.verticalAlignment = SWT.TOP;
+		data.grabExcessVerticalSpace = true;
+		description.setLayoutData(data);
 
 		initializeFields();
 		Dialog.applyDialogFont(composite);
-		verifyComplete();
 		return composite;
 	}
 
-	protected void verifyComplete() {
-		if (url.getText().trim().length() == 0) {
-			setValid(false);
-			setErrorMessage(ProvUIMessages.RepositoryGroup_URLRequired);
-		}
-		setValid(true);
-		setErrorMessage(null);
-	}
-
-	public boolean performOk() {
-		if (repository.isModifiable()) {
-			repository.setName(name.getText().trim());
-		}
-		return super.performOk();
-	}
-
 	private void initializeFields() {
-		if (repository == null) {
-			url.setText("http://"); //$NON-NLS-1$
-		} else {
-			url.setText(repository.getLocation().toExternalForm());
-			name.setText(repository.getName());
-		}
+		// Shouldn't happen since we checked this before creating any controls
+		if (repositoryElement == null)
+			return;
+		url.setText(repositoryElement.getLocation().toExternalForm());
+		name.setText(repositoryElement.getName());
+		description.setText(repositoryElement.getDescription());
 	}
 
-	protected IRepository getRepository() {
-		if (repository == null) {
-			repository = (IRepository) ProvUI.getAdapter(getElement(), IRepository.class);
+	protected RepositoryElement getRepositoryElement() {
+		if (repositoryElement == null) {
+			repositoryElement = (RepositoryElement) ProvUI.getAdapter(getElement(), RepositoryElement.class);
 		}
-		return repository;
+		return repositoryElement;
 	}
 }
diff --git a/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/p2/ui/model/MetadataRepositoryElement.java b/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/p2/ui/model/MetadataRepositoryElement.java
index 4496a14..83f9dd0 100644
--- a/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/p2/ui/model/MetadataRepositoryElement.java
+++ b/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/p2/ui/model/MetadataRepositoryElement.java
@@ -11,11 +11,13 @@
 package org.eclipse.equinox.p2.ui.model;
 
 import java.net.URL;
+import org.eclipse.core.runtime.IProgressMonitor;
 import org.eclipse.equinox.internal.p2.ui.ProvUIMessages;
 import org.eclipse.equinox.internal.p2.ui.model.RemoteQueriedElement;
 import org.eclipse.equinox.p2.core.ProvisionException;
 import org.eclipse.equinox.p2.core.repository.IRepository;
 import org.eclipse.equinox.p2.metadata.repository.IMetadataRepository;
+import org.eclipse.equinox.p2.metadata.repository.IMetadataRepositoryManager;
 import org.eclipse.equinox.p2.query.IQueryable;
 import org.eclipse.equinox.p2.ui.ProvUIImages;
 import org.eclipse.equinox.p2.ui.operations.ProvisioningUtil;
@@ -57,7 +59,7 @@ public class MetadataRepositoryElement extends RemoteQueriedElement implements R
 		if (name != null && name.length() > 0) {
 			return name;
 		}
-		return getURL().toExternalForm();
+		return getLocation().toExternalForm();
 	}
 
 	/*
@@ -67,12 +69,23 @@ public class MetadataRepositoryElement extends RemoteQueriedElement implements R
 	 */
 	public IQueryable getQueryable() {
 		if (queryable == null)
+			return getMetadataRepository(null);
+		return queryable;
+	}
+
+	public IRepository getRepository(IProgressMonitor monitor) {
+		return getMetadataRepository(monitor);
+	}
+
+	private IMetadataRepository getMetadataRepository(IProgressMonitor monitor) {
+		if (queryable == null)
 			try {
-				queryable = ProvisioningUtil.loadMetadataRepository(url, null);
+				queryable = ProvisioningUtil.loadMetadataRepository(url, monitor);
 			} catch (ProvisionException e) {
 				handleException(e, NLS.bind(ProvUIMessages.MetadataRepositoryElement_RepositoryLoadError, url));
 			}
-		return queryable;
+		return (IMetadataRepository) queryable;
+
 	}
 
 	/*
@@ -88,7 +101,7 @@ public class MetadataRepositoryElement extends RemoteQueriedElement implements R
 	/* (non-Javadoc)
 	 * @see org.eclipse.equinox.p2.ui.model.RepositoryElement#getURL()
 	 */
-	public URL getURL() {
+	public URL getLocation() {
 		return url;
 	}
 
@@ -98,7 +111,25 @@ public class MetadataRepositoryElement extends RemoteQueriedElement implements R
 	 */
 	public String getName() {
 		try {
-			return ProvisioningUtil.getMetadataRepositoryName(url);
+			String name = ProvisioningUtil.getMetadataRepositoryProperty(url, IMetadataRepositoryManager.PROP_NAME);
+			if (name == null)
+				return ""; //$NON-NLS-1$
+			return name;
+		} catch (ProvisionException e) {
+			return ""; //$NON-NLS-1$
+		}
+	}
+
+	/*
+	 * (non-Javadoc)
+	 * @see org.eclipse.equinox.p2.ui.model.RepositoryElement#getDescription()
+	 */
+	public String getDescription() {
+		try {
+			String description = ProvisioningUtil.getMetadataRepositoryProperty(url, IMetadataRepositoryManager.PROP_DESCRIPTION);
+			if (description == null)
+				return ""; //$NON-NLS-1$
+			return description;
 		} catch (ProvisionException e) {
 			return ""; //$NON-NLS-1$
 		}
diff --git a/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/p2/ui/model/RepositoryElement.java b/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/p2/ui/model/RepositoryElement.java
index dcefe7d..83a9ce1 100644
--- a/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/p2/ui/model/RepositoryElement.java
+++ b/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/p2/ui/model/RepositoryElement.java
@@ -11,6 +11,8 @@
 package org.eclipse.equinox.p2.ui.model;
 
 import java.net.URL;
+import org.eclipse.core.runtime.IProgressMonitor;
+import org.eclipse.equinox.p2.core.repository.IRepository;
 
 /**
  * Interface for elements that represent repositories.
@@ -19,7 +21,11 @@ import java.net.URL;
  */
 public interface RepositoryElement {
 
-	public URL getURL();
+	public URL getLocation();
 
 	public String getName();
+
+	public String getDescription();
+
+	public IRepository getRepository(IProgressMonitor monitor);
 }
diff --git a/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/p2/ui/operations/ProvisioningUtil.java b/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/p2/ui/operations/ProvisioningUtil.java
index 99fccc5..9249e8c 100644
--- a/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/p2/ui/operations/ProvisioningUtil.java
+++ b/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/p2/ui/operations/ProvisioningUtil.java
@@ -44,11 +44,11 @@ public class ProvisioningUtil {
 		ProvUIActivator.getDefault().notifyListeners(event);
 	}
 
-	public static String getMetadataRepositoryName(URL location) throws ProvisionException {
+	public static String getMetadataRepositoryProperty(URL location, String key) throws ProvisionException {
 		IMetadataRepositoryManager manager = (IMetadataRepositoryManager) ServiceHelper.getService(ProvUIActivator.getContext(), IMetadataRepositoryManager.class.getName());
 		if (manager == null)
 			throw new ProvisionException(ProvUIMessages.ProvisioningUtil_NoRepositoryManager);
-		return manager.getRepositoryProperty(location, IMetadataRepositoryManager.PROP_NAME);
+		return manager.getRepositoryProperty(location, key);
 	}
 
 	public static IMetadataRepository loadMetadataRepository(URL location, IProgressMonitor monitor) throws ProvisionException {
@@ -90,12 +90,12 @@ public class ProvisioningUtil {
 		ProvUIActivator.getDefault().notifyListeners(event);
 	}
 
-	public static String getArtifactRepositoryName(URL location) throws ProvisionException {
+	public static String getArtifactRepositoryProperty(URL location, String key) throws ProvisionException {
 		IArtifactRepositoryManager manager = (IArtifactRepositoryManager) ServiceHelper.getService(ProvUIActivator.getContext(), IArtifactRepositoryManager.class.getName());
 		if (manager == null) {
 			throw new ProvisionException(ProvUIMessages.ProvisioningUtil_NoRepositoryManager);
 		}
-		return manager.getRepositoryProperty(location, IArtifactRepositoryManager.PROP_NAME);
+		return manager.getRepositoryProperty(location, key);
 	}
 
 	public static IArtifactRepository loadArtifactRepository(URL location, IProgressMonitor monitor) throws ProvisionException {
diff --git a/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/p2/ui/viewers/ProvElementComparer.java b/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/p2/ui/viewers/ProvElementComparer.java
index d0cf738..01fc77a 100644
--- a/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/p2/ui/viewers/ProvElementComparer.java
+++ b/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/p2/ui/viewers/ProvElementComparer.java
@@ -64,7 +64,7 @@ public class ProvElementComparer implements IElementComparer {
 
 	private URL getRepositoryLocation(Object obj) {
 		if (obj instanceof RepositoryElement)
-			return ((RepositoryElement) obj).getURL();
+			return ((RepositoryElement) obj).getLocation();
 		return null;
 	}
 
